### PR TITLE
Result of a docstring formatter and some manual fixes on top of it

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |

## Description

Add a pre-commit hook to format our docstring automatically for pep257 using https://github.com/myint/docformatter

Possible options:
```
usage: docformatter [-h] [-i | -c] [-r] [--wrap-summaries length]
                    [--wrap-descriptions length] [--blank]
                    [--pre-summary-newline] [--make-summary-multi-line]
                    [--force-wrap] [--range line line] [--version]
                    files [files ...]

Formats docstrings to follow PEP 257.

positional arguments:
  files                 files to format or '-' for standard in

optional arguments:
  -h, --help            show this help message and exit
  -i, --in-place        make changes to files instead of printing diffs
  -c, --check           only check and report incorrectly formatted files
  -r, --recursive       drill down directories recursively
  -e, --exclude         exclude directories and files by names

  --wrap-summaries length
                        wrap long summary lines at this length; set to 0 to
                        disable wrapping (default: 79)
  --wrap-descriptions length
                        wrap descriptions at this length; set to 0 to disable
                        wrapping (default: 72)
  --blank               add blank line after description
  --pre-summary-newline
                        add a newline before the summary of a multi-line
                        docstring
  --make-summary-multi-line
                        add a newline before and after the summary of a one-
                        line docstring
  --force-wrap          force descriptions to be wrapped even if it may result
                        in a mess
  --range line line     apply docformatter to docstrings between these lines;
                        line numbers are indexed at 1
  --version             show program's version number and exit
```  

Suggested by @DanielNoord https://github.com/PyCQA/pylint/pull/5620#discussion_r777000979

### CODE REVIEW

File: doc/exts/pylint_extensions.py [hunk 1]
```diff
@@ -2,7 +2,8 @@
 # Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 # For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
 
-"""Script used to generate the extensions file before building the actual documentation."""
+"""Script used to generate the extensions file before building the actual
+documentation."""
```
Comment (DanielNoord) [inline doc/exts/pylint_extensions.py:6]: This doesn't actually enforce PEP8. See [here](https://www.python.org/dev/peps/pep-0008/#id31).
Perhaps there is a setting to toggle, but I would like a checker that at least does this 😄

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Not sure if this is the right tool for the job.

I took a quick look through the repository and some open issues can also be problematic for use.
For one there is:
https://github.com/myint/docformatter/issues/9 with associated PR https://github.com/myint/docformatter/pull/49.
It has been open since 2015 and I feel like it is one of the conventions we would like to enforce.

Another is:
https://github.com/myint/docformatter/issues/75

I'd like to find a good tool to do this, but this one might not be it sadly 😢

File: pylint/checkers/similar.py [hunk 2]
```diff
@@ -403,13 +398,13 @@ def append_stream(
...
     def run(self) -> None:
-        """start looking for similarities and display results on stdout"""
+        """start looking for similarities and display results on stdout."""
         if self.min_lines == 0:
             return
         self._display_sims(self._compute_sims())
 
     def _compute_sims(self) -> List[Tuple[int, Set[LinesChunkLimits_T]]]:
-        """compute similarities in appended files"""
+        """compute similarities in appended files."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:407]:
```suggestion
        """Compute similarities in appended files."""
```

File: pylint/checkers/similar.py [hunk 3]
```diff
@@ -544,25 +539,24 @@ def _find_common(
                 yield com
 
     def _iter_sims(self) -> Generator[Commonality, None, None]:
-        """iterate on similarities among all files, by making a cartesian
-        product
-        """
+        """iterate on similarities among all files, by making a cartesian product."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:542]:
```suggestion
        """Iterate on similarities among all files, by making a Cartesian product."""
```

File: pylint/checkers/similar.py [hunk 4]
```diff
@@ -730,17 +725,18 @@ def report_similarities(
     stats: LinterStats,
     old_stats: Optional[LinterStats],
 ) -> None:
-    """make a layout with some stats about duplication"""
+    """make a layout with some stats about duplication."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:728]:
```suggestion
    """Make a layout with some stats about duplication."""
```

File: pylint/checkers/similar.py [hunk 5]
```diff
@@ -730,17 +725,18 @@ def report_similarities(
...
     lines += table_lines_from_stats(stats, old_stats, "duplicated_lines")
     sect.append(Table(children=lines, cols=4, rheaders=1, cheaders=1))
 
 
 # wrapper to get a pylint checker from the similar class
 class SimilarChecker(BaseChecker, Similar, MapReduceMixin):
-    """checks for similarities and duplicated code. This computation may be
-    memory / CPU intensive, so you should disable it if you experiment some
-    problems.
+    """checks for similarities and duplicated code.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:736]:
```suggestion
    """Checks for similarities and duplicated code.
```

File: pylint/checkers/similar.py [hunk 6]
```diff
@@ -829,12 +825,12 @@ def set_option(self, optname, value, action=None, optdict=None):
             self.ignore_signatures = self.config.ignore_signatures
 
     def open(self):
-        """init the checkers: reset linesets and statistics information"""
+        """init the checkers: reset linesets and statistics information."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:828]:
```suggestion
        """Init the checkers: reset linesets and statistics information."""
```

File: pylint/checkers/similar.py [hunk 7]
```diff
@@ -829,12 +825,12 @@ def set_option(self, optname, value, action=None, optdict=None):
...
         self.linesets = []
         self.linter.stats.reset_duplicated_lines()
 
     def process_module(self, node: nodes.Module) -> None:
-        """process a module
+        """process a module.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:833]:
```suggestion
        """Process a module.
```

File: pylint/checkers/similar.py [hunk 8]
```diff
@@ -888,7 +885,7 @@ def register(linter: "PyLinter") -> None:
 
 
 def usage(status=0):
-    """display command line usage information"""
+    """display command line usage information."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:888]:
```suggestion
    """Display command line usage information."""
```

File: pylint/checkers/similar.py [hunk 9]
```diff
@@ -899,7 +896,7 @@ def usage(status=0):
 
 
 def Run(argv=None):
-    """standalone command line access point"""
+    """standalone command line access point."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/similar.py:899]:
```suggestion
    """Standalone command line access point."""
```

File: pylint/checkers/spelling.py [hunk 10]
```diff
@@ -461,7 +459,7 @@ def visit_functiondef(self, node: nodes.FunctionDef) -> None:
     visit_asyncfunctiondef = visit_functiondef
 
     def _check_docstring(self, node):
-        """check the node has any spelling errors"""
+        """check the node has any spelling errors."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/spelling.py:462]:
```suggestion
        """Check the node has any spelling errors."""
```

File: pylint/checkers/typecheck.py [hunk 11]
```diff
@@ -55,8 +55,7 @@
 # Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 # For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
 
-"""try to find more bugs in the code using astroid inference capabilities
-"""
+"""try to find more bugs in the code using astroid inference capabilities."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:58]:
```suggestion
"""Try to find more bugs in the code using astroid inference capabilities."""
```

File: pylint/checkers/typecheck.py [hunk 12]
```diff
@@ -782,7 +779,7 @@ def _is_invalid_isinstance_type(arg):
 
 
 class TypeChecker(BaseChecker):
-    """try to find bugs in the code using type inference"""
+    """try to find bugs in the code using type inference."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:782]:
```suggestion
    """Try to find bugs in the code using type inference."""
```

File: pylint/checkers/typecheck.py [hunk 13]
```diff
@@ -1007,7 +1004,7 @@ def visit_delattr(self, node: nodes.DelAttr) -> None:
 
     @check_messages("no-member", "c-extension-no-member")
     def visit_attribute(self, node: nodes.Attribute) -> None:
-        """check that the accessed attribute exists
+        """check that the accessed attribute exists.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1007]:
```suggestion
        """Check that the accessed attribute exists.
```

File: pylint/checkers/typecheck.py [hunk 14]
```diff
@@ -1137,17 +1134,14 @@ def _get_nomember_msgid_hint(self, node, owner):
...
+        """Process assignments in the AST."""
 
         self._check_assignment_from_function_call(node)
         self._check_dundername_is_string(node)
 
     def _check_assignment_from_function_call(self, node):
-        """check that if assigning to a function call, the function is
-        possibly returning something valuable
-        """
+        """check that if assigning to a function call, the function is possibly
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1143]:
```suggestion
        """Check that if assigning to a function call, the function is possibly
```

File: pylint/checkers/typecheck.py [hunk 15]
```diff
@@ -1317,10 +1308,9 @@ def _check_isinstance_args(self, node):
     # pylint: disable=too-many-branches,too-many-locals
     @check_messages(*(list(MSGS.keys())))
     def visit_call(self, node: nodes.Call) -> None:
-        """check that called functions/methods are inferred to callable objects,
-        and that the arguments passed to the function match the parameters in
-        the inferred function's definition
-        """
+        """check that called functions/methods are inferred to callable objects, and
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1311]:
```suggestion
        """Check that called functions/methods are inferred to callable objects, and
```

File: pylint/checkers/utils.py [hunk 16]
```diff
@@ -57,8 +57,7 @@
 # Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 # For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
 
-"""some functions that may be useful for various checkers
-"""
+"""some functions that may be useful for various checkers."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:60]:
```suggestion
"""Some functions that may be useful for various checkers."""
```

File: pylint/checkers/utils.py [hunk 17]
```diff
@@ -315,14 +314,14 @@ def get_all_elements(
 
 
 def is_super(node: nodes.NodeNG) -> bool:
-    """return True if the node is referencing the "super" builtin function"""
+    """return True if the node is referencing the "super" builtin function."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:317]:
```suggestion
    """Return True if the node is referencing the "super" builtin function."""
```

File: pylint/checkers/utils.py [hunk 18]
```diff
@@ -511,10 +507,11 @@ def __init__(self, index):
...
 ) -> Tuple[Set[str], int, Dict[str, str], List[str]]:
-    """Parses a format string, returning a tuple of (keys, num_args), where 'keys'
-    is the set of mapping keys in the format string, and 'num_args' is the number
-    of arguments required by the format string. Raises IncompleteFormatString or
-    UnsupportedFormatCharacter if a parse error occurs.
+    """Parses a format string, returning a tuple of (keys, num_args), where 'keys' is
+    the set of mapping keys in the format string, and 'num_args' is the number of
+    arguments required by the format string.
+
+    Raises IncompleteFormatString or UnsupportedFormatCharacter if a parse error occurs.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:514]:
```suggestion
    """Parses a format string, returning a tuple of (keys, num_args).
    
     Where 'keys' is
    the set of mapping keys in the format string, and 'num_args' is the number of
    arguments required by the format string.
    
    Raises IncompleteFormatString or UnsupportedFormatCharacter if a parse error occurs.
```

File: pylint/checkers/variables.py [hunk 19]
```diff
@@ -187,7 +186,7 @@ def in_for_else_branch(parent, stmt):
 
 @lru_cache(maxsize=1000)
 def overridden_method(klass, name):
-    """get overridden method if any"""
+    """get overridden method if any."""
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:189]:
```suggestion
    """Get overridden method if any."""
```

File: pylint/checkers/variables.py [hunk 20]
```diff
@@ -584,12 +578,12 @@ def consumed(self):
...
-        have executed, such as statements in except blocks, or statements
-        in try blocks (when evaluating their corresponding except and finally
-        blocks). Checkers that want to treat the statements as executed
-        (e.g. for unused-variable) may need to add them back.
+        """Retrieves nodes filtered out by get_next_to_consume() that may not have
+        executed, such as statements in except blocks, or statements in try blocks (when
+        evaluating their corresponding except and finally blocks).
+
+        Checkers that want to treat the statements as executed (e.g. for unused-
+        variable) may need to add them back.
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:586]:
```suggestion
        """Retrieves nodes filtered out by get_next_to_consume()
        
        Those nodes may not have        executed, such as statements in except blocks, or statements in try blocks (when
        evaluating their corresponding except and finally blocks).

        Checkers that want to treat the statements as executed (e.g. for unused-
        variable) may need to add them back.
```

File: pylint/checkers/variables.py [hunk 21]
```diff
@@ -1834,8 +1825,9 @@ def _is_first_level_self_reference(
     def _is_never_evaluated(
         defnode: nodes.NamedExpr, defnode_parent: nodes.IfExp
     ) -> bool:
-        """Check if a NamedExpr is inside a side of if ... else that never
-        gets evaluated
+        """Check if a NamedExpr is inside a side of if ...
+
+        else that never gets evaluated
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/variables.py:1830]:
```suggestion
        """Check if a NamedExpr is inside of an 'if else' that never gets evaluated
```

File: pylint/extensions/typing.py [hunk 22]
```diff
@@ -257,9 +257,10 @@ def _check_for_typing_alias(
 
     @check_messages("consider-using-alias")
     def leave_module(self, node: nodes.Module) -> None:
-        """After parsing of module is complete, add messages for
-        'consider-using-alias' check. Make sure results are safe
-        to recommend / collision free.
+        """After parsing of module is complete, add messages for 'consider- using-alias'
+        check.
```
Comment (Pierre-Sassoulas) [inline pylint/extensions/typing.py:261]:
```suggestion
        """After the parsing of a module add messages for 'consider-using-alias'.
```

File: doc/exts/pylint_extensions.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply doc/exts/pylint_extensions.py:6]: I'm going to apply the good modification for a start, and maybe fix the tool at this source (temporary fork ?). It's the only autofixer I found.

Comment (DanielNoord) [reply doc/exts/pylint_extensions.py:6]: Tools seems to be making quite a bit of unnecessary and arbitrary changes.
Somebody should probably take a good look at all suggested changes.

I would not put the tool in `pre-commit` for now. It doesn't really do what we're looking for.
Perhaps I'll work on something like this a hobby project 😄

Comment (Pierre-Sassoulas) [reply doc/exts/pylint_extensions.py:6]: Sure it's not usable in the current form. There are some strange options. One to always line break before and after string, one only before. I guess I could fix that to line break when a single line would not work. Also I'd like to add capitalization for the first letter, probably not that hard. All in all I think the tool is fixable and I'm going to try it next year 😄 .

File: doc/exts/pylint_features.py [hunk 23]
```diff
@@ -2,7 +2,8 @@
 # Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 # For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
 
-"""Script used to generate the features file before building the actual documentation."""
+"""Script used to generate the features file before building the actual
+documentation."""
```
Comment (cdce8p) [inline doc/exts/pylint_features.py:6]: Please let's not do that. It's just not worth it. Not everything needs to be perfect.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I have started working on a little tool to do some of the things that `docformatter` does.

I think it should be ready for a `0.1.0` somewhere in the coming weeks. If we want to go ahead with a docstring auto formatter we could use it as we'll have more influence on what it changes.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas The latest fix also seems to have affected functional tests? Was this your intention? Are these 6k changes all manual?

**Comment by Pierre-Sassoulas** [comment]
> Are these 6k changes all manual?

It's docformatter + manual and semi manual (sed) after that. I should not have touched the functional tests. I'm going to fix that, but I think doing https://github.com/DanielNoord/pydocstringformatter/issues/3 and https://github.com/DanielNoord/pydocstringformatter/issues/5 before would be cleaner.

**Comment by DanielNoord** [comment]
Shall we convert this to draft while I work on `pydocstringformatter` a bit more? You're more than welcome to work on them as well! I'll make you contributor on the repository so you can assign yourself and we don't do double work!

**Comment by Pierre-Sassoulas** [comment]
Closing in favor of #5910

**Pull request closed without merging**